packages/agents/src/mcp/client.ts

```diff
@@ -55,173 +98,340 @@ export class MCPClientManager {
+          authProvider
+        },
+        authUrl: server.auth_url ?? undefined,
+        clientId: server.client_id ?? undefined
+      });
+
+      if (needsOAuth) {
+        // OAuth server - just set state to authenticating (wait for OAuth flow)
+        if (this.mcpConnections[server.id]) {
+          this.mcpConnections[server.id].connectionState = "authenticating";
+        }
+      } else {
+        // Non-OAuth server - connect immediately
+        await this.connectToServer(server.id).catch((error) => {
+          console.error(`Error restoring ${server.id}:`, error);
+        });
+      }
     }
 
-    // Initialize connection first
-    await this.mcpConnections[id].init();
+    this._isRestored = true;
+  }
 
-    // Handle OAuth completion if we have a reconnect code
-    if (options.reconnect?.oauthCode) {
+  /**
+   * Register an MCP server connection without connecting
+   * Creates the connection object, sets up observability, and saves to storage
+   *
+   * @param id Server ID
+   * @param options Registration options including URL, name, callback URL, and connection config
+   * @returns Server ID
+   */
+  registerServer(id: string, options: RegisterServerOptions): string {
+    // Skip if connection already exists
+    if (this.mcpConnections[id]) {
+      return id;
+    }
+
+    const normalizedTransport = {
+      ...options.transport,
+      type: options.transport?.type ?? ("auto" as TransportType)
+    };
+
+    this.mcpConnections[id] = new MCPClientConnection(
+      new URL(options.url),
+      {
+        name: this._name,
+        version: this._version
+      },
+      {
+        client: options.client ?? {},
+        transport: normalizedTransport
+      }
+    );
+
+    // Pipe connection-level observability events to the manager-level emitter
+    const store = new DisposableStore();
+    const existing = this._connectionDisposables.get(id);
+    if (existing) existing.dispose();
+    this._connectionDisposables.set(id, store);
+    store.add(
+      this.mcpConnections[id].onObservabilityEvent((event) => {
+        this._onObservabilityEvent.fire(event);
+      })
+    );
+
+    // Save to storage
+    this._storage.saveServer({
+      id,
+      name: options.name,
+      server_url: options.url,
+      callback_url: options.callbackUrl,
+      client_id: options.clientId ?? null,
+      auth_url: options.authUrl ?? null,
+      server_options: JSON.stringify({
+        client: options.client,
+        transport: options.transport
+      })
+    });
+
+    return id;
+  }
+
+  /**
+   * Connect to an already registered MCP server
+   * Updates storage with auth URL and client ID after connection
+   *
+   * @param id Server ID
+   * @param options Connection options (e.g., OAuth code for completing OAuth flow)
+   * @returns Auth URL if OAuth is required, undefined otherwise
+   */
+  async connectToServer(
+    id: string,
+    options?: ConnectServerOptions
+  ): Promise<{
+    authUrl?: string;
```

Comment:
this signature is kinda dumb. we need to revisit why we even did this, and whether there's a better way to extract it